crates/pathfinder/src/storage/state.rs

```diff
@@ -1047,6 +1062,80 @@ impl ContractsStateTable {
     }
 }
 
+/// Stores all known [Starknet state updates][crate::rpc::types::reply::StateUpdate].
+pub struct StarknetStateUpdatesTable {}
+impl StarknetStateUpdatesTable {
+    /// Inserts a StarkNet state update accociated with a particular block into the [StarknetStateUpdatesTable].
+    ///
+    /// Overwrites existing data if the block hash already exists.
+    pub fn insert(
+        tx: &Transaction<'_>,
+        block_hash: StarknetBlockHash,
+        state_update: &StateUpdate,
+    ) -> anyhow::Result<()> {
+        let serialized =
+            serde_json::to_vec(&state_update).context("Serialize Starknet state update")?;
+
+        let mut compressor = zstd::bulk::Compressor::new(10).context("Create zstd compressor")?;
+        let compressed = compressor
+            .compress(&serialized)
+            .context("Compress Starknet state update")?;
+
+        tx.execute(
+            r"INSERT INTO starknet_state_updates (block_hash, data) VALUES (:block_hash, :data)",
+            named_params![
+                ":block_hash": block_hash.0.as_be_bytes(),
+                ":data": &compressed,
+            ],
+        )
+        .context("Insert state update data into state updates table")?;
+
+        Ok(())
+    }
+
+    /// Gets a StarkNet state update for block.
+    pub fn get(
+        tx: &Transaction<'_>,
+        block_hash: StarknetBlockHash,
+    ) -> anyhow::Result<Option<StateUpdate>> {
+        let mut stmt = tx
+            .prepare("SELECT data FROM starknet_state_updates WHERE block_hash = ?1")
+            .context("Preparing statement")?;
+
+        let mut rows = stmt
+            .query(params![block_hash.0.as_be_bytes()])
+            .context("Executing query")?;
+
+        let row = match rows.next()? {
+            Some(row) => row,
+            None => return Ok(None),
+        };
+
+        let state_update = row.get_ref_unwrap(0).as_blob()?;
+        let state_update = zstd::decode_all(state_update).context("Decompressing state update")?;
+        let state_update =
+            serde_json::from_slice(&state_update).context("Deserializing state update")?;
+
+        Ok(Some(state_update))
+    }
+
+    /// Deletes all rows from __head down-to reorg_tail__
+    /// i.e. it deletes all rows where `block number >= reorg_tail`.
+    pub fn reorg(tx: &Transaction<'_>, reorg_tail: StarknetBlockNumber) -> anyhow::Result<()> {
+        tx.execute(
+            r"DELETE FROM starknet_state_updates
+            WHERE ROWID IN (
+                SELECT updates.ROWID FROM starknet_state_updates updates
+                INNER JOIN starknet_blocks blocks
+                    ON (updates.block_hash = blocks.hash)
+                WHERE blocks.number >= ?
+            );",
+            params![reorg_tail.0],
+        )?;
+        Ok(())
+    }
```

Comment:
Makes sense, I don't mind being able to serve state updates which were reorged-away.

Reply:
I could leave the reorg functionality in case we change our mind but actually not call it in the sync process. Does it make sense?

Reply:
Following a discussion with @kkovaacs we decided to be consistent with how the block data is handled during a reorg. This is to avoid further complications when the sync process is refactored to allow many concurrent data sources (forks) which will definitely lead to a major schema restructuring and migration.

Having state updates that refer to block hashes for whose blocks in the DB don't exist would then have to be accounted for to make sure the above migration works properly.